graspy/cluster/autogmm.py

```diff
@@ -0,0 +1,599 @@
+                "affinity",
+                "linkage",
+                "covariance_type",
+                "reg_covar",
+            ]
+        )
+
+        for params in param_grid:
+            if label_init is not None:
+                onehot = _labels_to_onehot(label_init)
+                weights_init, means_init, precisions_init = _onehot_to_initial_params(
+                    X, onehot, params[1]["covariance_type"]
+                )
+                gm_params = params[1]
+                gm_params["weights_init"] = weights_init
+                gm_params["means_init"] = means_init
+                gm_params["precisions_init"] = precisions_init
+            elif params[0]["affinity"] != "none":
+                agg = AgglomerativeClustering(**params[0])
+                n = X.shape[0]
+
+                if n > self.max_agglom_size:  # if dataset is huge, agglomerate a subset
+                    subset_idxs = np.random.choice(
+                        np.arange(0, n), self.max_agglom_size
+                    )
+                    X_subset = X[subset_idxs, :]
+                else:
+                    X_subset = X
+                agg_clustering = agg.fit_predict(X_subset)
+                onehot = _labels_to_onehot(agg_clustering)
+                weights_init, means_init, precisions_init = _onehot_to_initial_params(
+                    X_subset, onehot, params[1]["covariance_type"]
+                )
+                gm_params = params[1]
+                gm_params["weights_init"] = weights_init
+                gm_params["means_init"] = means_init
+                gm_params["precisions_init"] = precisions_init
+            else:
+                gm_params = params[1]
+                gm_params["init_params"] = "kmeans"
+            gm_params["reg_covar"] = 0
+            gm_params["max_iter"] = max_iter
+            gm_params["verbose"] = verbose
+
+            bic = np.inf  # if none of the iterations converge, bic is set to inf
+            # below is the regularization scheme
+            while gm_params["reg_covar"] <= 1:
+                model = GaussianMixture(**gm_params)
+                try:
+                    model.fit(X)
+                    predictions = model.predict(X)
+                    counts = [
+                        sum(predictions == i) for i in range(gm_params["n_components"])
+                    ]
+                    # singleton clusters not allowed
+                    assert not any([count <= 1 for count in counts])
+
+                except ValueError:
+                    gm_params["reg_covar"] = _increase_reg(gm_params["reg_covar"])
+                    continue
+                except AssertionError:
+                    gm_params["reg_covar"] = _increase_reg(gm_params["reg_covar"])
+                    continue
+                # if the code gets here, then the model has been fit with no errors or
+                # singleton clusters
+                bic = model.bic(X)
```

Comment:
i wonder if this would be more clear by adding `and bic == np.inf` as a condition for the while, thoughts?

Reply:
good call, in fact I think you caught an algorithmic mistake